src/main/cpp/src/timezones.cu

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <thrust/binary_search.h>
+
+#include "timezones.hpp"
+
+using column = cudf::column;
+using column_device_view = cudf::column_device_view;
+using lists_column_device_view = cudf::detail::lists_column_device_view;
+using size_type = cudf::size_type;
+using struct_view = cudf::struct_view;
+
+namespace {
+
+/**
+ * @brief Convert the timestamp value to either UTC or a specified timezone
+ *
+ * This device function uses a binary search to find the instant of the transition
+ * to find the right offset to do the transition.
+ *
+ * To transition to UTC: do a binary search on the tzInstant child column and subtract
+ * the offset
+ *
+ * To transition from UTC: do a binary search on the utcInstant child column and add
+ * the offset
+ *
+ * @tparam typestamp_type type of the input and output timestamp
+ * @param timestamp input timestamp
+ * @param transitions the list column of transitions to figure out the correct offset
+ *                    to adjust the timestamp. The type of the values in this column is
+ *                    LIST<STRUCT<utcInstant: int64, tzInstant: int64, utcOffset: int32>>
+ * @param tz_index the index of the specified zone id in the transitions table
+ * @param to_utc whether we are converting to UTC or converting to the timezone
+ */
+template <typename timestamp_type>
+__device__ timestamp_type convert_timestamp_timezone(timestamp_type const &timestamp,
+                                                     lists_column_device_view const &transitions,
+                                                     size_type tz_index, bool to_utc) {
+
+  using duration_type = typename timestamp_type::duration;
+  using cuda::std::chrono::duration_cast;
```

Comment:
Please do not type alias any name having `chrono::` prefix. It is very error-prone.

Reply:
Why is this specifically error prone?